### PR TITLE
Introducing the golang plugin

### PR DESCRIPTION
**Pull request by apahim**
golang ships a builtin test runner, which is called by
`go test <opts> <package or path>`. This patch introduces the golang
optional plugin, which can be used to discover and run golang tests
using Avocado.

Signed-off-by: Amador Pahim <apahim@redhat.com>

### CODE REVIEW

**Comment by apahim** [comment]
[Introduce Golang pluign](https://trello.com/c/velCpIvk/1143-introduce-golang-pluign)

File: docs/source/optional_plugins/golang.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
+=============
+Golang Plugin
+=============
+
+This optional plugin enables Avocado to work with tests originally
```
Comment (clebergnu) [inline docs/source/optional_plugins/golang.rst]: Suggestion:

"This optional plugin enables Avocado to list and run tests written using ..."

The idea being that "originally", to me, sounds like tests have to change to work with Avocado.

File: docs/source/optional_plugins/golang.rst [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+written using the `Go testing package <https://golang.org/pkg/testing/>`_.
+
+To install the Golang plugin from pip, use::
+
+    $ sudo pip install avocado-framework-plugin-golang
+
+After installed, you can list/run Golang tests providing the package name::
+
+    ~$ avocado list golang.org/x/text/unicode/norm
```
Comment (clebergnu) [inline docs/source/optional_plugins/golang.rst]: While running with an invalid reference:

```
$ avocado list /path/doesnot/exist
Test discovery plugin <avocado_golang.GolangLoader object at 0x7f4a6be46e90> failed: 'NoneType' object has no attribute 'endswith'
Unable to resolve reference(s) '/path/doesnot/exist' with plugins(s) 'file', 'golang', 'yaml_testsuite', 'robot', 'external', try running 'avocado list -V /path/doesnot/exist' to see the details.
```

The first line is most probably not a good idea.

File: optional_plugins/golang/avocado_golang/__init__.py [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+    def __init__(self, args, extra_params):
+        super(GolangLoader, self).__init__(args, extra_params)
+
+    def discover(self, url, which_tests=loader.DEFAULT):
+        avocado_suite = []
+        test_files = []
+        _subtest = None
+        tests_filter = None
+        go_src = os.path.join(os.environ.get('GOPATH'), 'src')
```
Comment (clebergnu) [inline optional_plugins/golang/avocado_golang/__init__.py]: Maybe this should be something like:

```python
go_src = os.path.join(os.environ.get('GOPATH', os.path.expanduser('~/go'), 'src'))
```

From what I read, `~/go` is the default location.

**Comment by apahim** [comment]
@clebergnu, yes, my fault. `GOROOT` is not being checked in this version. It will be present in V2. Thanks.

File: optional_plugins/golang/avocado_golang/__init__.py (hunk 3 above)
Comment (apahim) [reply optional_plugins/golang/avocado_golang/__init__.py]: Ack

File: docs/source/optional_plugins/golang.rst (hunk 2 above)
Comment (apahim) [reply docs/source/optional_plugins/golang.rst]: Ops :)

File: docs/source/optional_plugins/golang.rst (hunk 1 above)
Comment (apahim) [reply docs/source/optional_plugins/golang.rst]: Ok, good.

**Pull request closed without merging**